transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,162 @@
+ * buffer are removed the promise associated with it will be completed when promise passed in during removal of the
+ * last byte completes.
+ *
+ * <p>This functionality is useful for aggregating or partitioning writes into fixed size buffers for framing protocols
+ * such as HTTP2.
+ */
+public class CoalescingBufferQueue {
+
+    private final ByteBufAllocator alloc;
+    private final ArrayDeque<Object> promisesAndBufs = new ArrayDeque<Object>();
+    private int readableBytes;
+
+    public CoalescingBufferQueue(ByteBufAllocator alloc) {
+        this.alloc = alloc;
+    }
+
+    /**
+     * Add a buffer to the end of the queue.
+     */
+    public void add(ByteBuf buf) {
+        promisesAndBufs.add(ObjectUtil.checkNotNull(buf, "buffer"));
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all it's bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        promisesAndBufs.add(ObjectUtil.checkNotNull(buf, "buffer"));
+        if (!ObjectUtil.checkNotNull(promise, "promise").isVoid()) {
+            promisesAndBufs.add(promise);
+        }
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Remove a buffer from the queue with the specified number of bytes. Any added buffer who's bytes are fully
+     * consumed during removal will have it's promise completed when the aggregatePromise completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}.
+     * @param aggregatePromise used to aggregate the promises for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
+        // Use isEmpty rather than readableBytes==0 as we may have a promise associated with an empty buffer.
+        if (promisesAndBufs.isEmpty()) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
+        CompositeByteBuf composite = null;
+        int originalBytes = bytes;
+        while (!promisesAndBufs.isEmpty()) {
+            Object entry = promisesAndBufs.remove();
+            if (entry instanceof ChannelPromise) {
+                aggregatePromise.addListener(new ChannelPromiseNotifier((ChannelPromise) entry));
+            } else if (bytes > 0) {
+                ByteBuf buf = (ByteBuf) entry;
+                if (toReturn == null) {
+                    if (buf.readableBytes() > bytes) {
```

Comment:
store buf.readableBytes() in a local var and reuse it. So we not need to call it multiple times if the first if not matches.